platform/engineering/design-docs/2021-06-11-small-batch-publishing.md

```diff
@@ -0,0 +1,350 @@
+
+The Publish Batch entities can also be used as an audit trail.
+
+Publish Batch Format :
+
+*   Batch UUID 
+*   Trigger {nid, vid, path} 
+*   deps [{nid, vid, path}]
+*   advanced queue integration?
+*   state 
+    *   Queued 
+    *   In Progress 
+    *   Complete 
+    *   Error
+*   Messages
+
+Example JSON message:
+
+```
+{
+  "uuid": "xxxx-33333-ffffd-4444"
+  "triggeredEntity": 
+    {
+      "nid": 123,
+      "vid": 2222,
+      "path" "path/on/va/website"
+    }
+  "updatedEntities: [
+    {
+      "nid": 1111,
+      "vid": 22222,
+      "path": "publish/path"
+    },
+    {
+      "nid": 33333,
+      "vid": 44444,
+      "path": "another/publish/path"
+    }
+  ],
+  "state": "IN_PROGRESS",
+  "messages": "This is a status message to gather error details or to communicate something to the editors"
+}
+```
+    
+
+#### Continuous Integration
+
+As part of continuous integration (CI), a Github Action will zip the function source and upload it to AWS in S3.
+
+*   Build the function and zip it.
+*   Get the checksum or hash of the ZIP file.
+*   Store the ZIP file in an S3 bucket at a key corresponding to the checksum if it doesn't already exist.
+
+The S3 key will be vetsgov-website-builds-s3-upload/single-page-build/master/function.zip on the master branch. Non-master branches will replace master in that key with the checksum of the ZIP file.
+
+Check the [list of function aliases](https://docs.aws.amazon.com/lambda/latest/dg/API_ListAliases.html) for an alias matching the checksum. If an alias doesn't exist yet for the checksum:
+
+*   [Update and publish the function code](https://docs.aws.amazon.com/lambda/latest/dg/API_UpdateFunctionCode.html) using the archived function from S3 for dev and staging lambda functions.
+*   [Create the alias](https://docs.aws.amazon.com/lambda/latest/dg/API_CreateAlias.html) for the new version of the function.
+
+If the current branch is `master`, also get the version that the `master` alias points to. If the master alias doesn't match the version of the checksum alias:
+
+*   [Update the master alias](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/lambda/update-alias.html) to the version currently aliased with the checksum.
+
+A second Github Action will be created that is [manually triggered](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/). The action will [Update and publish the function code](https://docs.aws.amazon.com/lambda/latest/dg/API_UpdateFunctionCode.html) the production lambda function.
+
+
+#### Assets
+
+Assets (images, pdfs) which are generated in Drupal will be uploaded to S3.  The content build will be updated to use the new S3 location as the `src` for the assets.  The details of this implementation will be in a different design doc and implemented outside of Small Batch Publishing.
```

Comment:
> The content build will be updated to use the new S3 location as the `src` for the assets.

Ideally we'd hope that the `src` are already set to S3 before the content build starts so we don't have to parse it. Is that possible?

Reply:
Yeah.  Initially, we were thinking of adding a new field to the Drupal file entities which contained the full S3 path.  Then we update the build/templates to use the new field.  This would remove any processing in the build.  We discussed this idea with @alexandec earlier this week.

Reply:
I added more detail here and linked to the epic.